apache/incubator-kie-tools
# KOGITO-6859: [DMN Editor] Enabling Import Java Classes feature

JIRA: https://issues.redhat.com/browse/KOGITO-6859

This PR enables a _first_ working version of the "Enabling Import Java Classes" feature.

**Changes introduced in this PR**

- JavaCodeCompletion service: The service API has been fixed. The reason was the `DelegateHandler` `executeCommand` methods returned a `CompletableFuture<?>` instead of a simple `Object` class. With this change applied, the service works without the previously reported exception.
- JavaCodeCompletion service: Updated pom dependencies and removed `gson` which is already imported via `tycho`
- JavaCodeCompletion service: Accessors retrieved in `GetAccessorsHandler` returns Simple name java classes as Type, while an FQCN is required for our scope. (eg. Prior to this PR, returned type was `Book`, instead of the required `com.package.Book`.
- DMN Editor: The "Import Java Classes" button is now shown in the editor, in the Data Type tab.
- ImportJavaClasses react component: Updates to improve the integration with the service, in particular for the Fields service. Some accessors need to be filtered out (eg. `getClass()`, `serialVersionUID`, the name reshaped (eg. remove `get` prefix from getters method`)
- ImportJavaClasses react component: Improved Tooltip management in the search box 
- ImportJavaClasses react component: Tests have been restored

**What is missing (next steps)**
- Lists and Arrays are currently not supported. I will manage it in a future PR.
- The `@KieActivator` annotation requires an improvement, as not a realistic solution in the current status (more details in the `How to Test` section.

**How to test it**
It requires:
- A Generated `vsix` file with the changes introduced here
- [Language Support for Java(TM) by Red Hat vscode plugin](https://marketplace.visualstudio.com/items?itemName=redhat.java)
- A project with Java classes
- A Java class marked with `@KieActivator` annotation. Unfortunatetly, at current status, we need to declare the annotation in our project. As mentioned in `What is missing` section, this is **temporary** and will be improved in a future PR. 
Steps are: 
1. Create a `org.kie.api.project` package in your project 
2. Create a `KieActivator` class under that package
3. Copy and paste this class declaration on it:
```
package org.kie.api.project;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import javax.inject.Qualifier;

import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Qualifier
@Target({ TYPE })
@Retention(RUNTIME)
public @interface KieActivator {

}
```
As a result, you should see this:

![Screenshot from 2022-04-08 14-32-11](https://user-images.githubusercontent.com/16005046/162436619-0c80206e-574d-4974-81f8-b6c35f08093e.png)

4. Add the `@KieActivator` annotation in one Java Class (eg. here I created a `Test.java` marked with the annotation)
![Screenshot from 2022-04-08 14-37-35](https://user-images.githubusercontent.com/16005046/162437000-dc7f1f42-636d-4edb-a42e-b88e96fff28f.png)

5. Now you are ready to use the feature. As shown in the below video

![Screencast-from-04-08-2022-03_40_12-PM](https://user-images.githubusercontent.com/16005046/162449909-170cf330-642f-424f-a36b-c38a92eecd9c.gif)
